File: 1996-the-number-of-weak-characters-in-the-game/1996-the-number-of-weak-characters-in-the-game.cpp
class Solution {
public:
    // We are checking that for each element , there exist a Next greater elemenet
    
    //(Increased defence value is existing or not)
    
    // If existing then include in our answer
    
    static bool compare(vector<int> &a , vector<int> &b)
    {
        if(a[0]==b[0])
        {
            return a[1]>b[1];
        }
        
        return a[0]<b[0];
    }
    int numberOfWeakCharacters(vector<vector<int>>& nums) {
        
        int n=nums.size();
        
        // See this is a greedy question 
        
        // sorted in terms of starting value , & if starting value is same
        
        // then we will put the max_end value first
        
        sort(nums.begin() , nums.end() , compare);
     
        vector<int> nge(n , n); // Find next greater element of the curr_val (defence as attack is already sorted )
        
        stack<int> stk;
        
        for(int i=nums.size()-1;i>=0;i--)
        {
            while(!stk.empty() && (nums[stk.top()][1]<=nums[i][1]))
            {
                stk.pop();
            }
            
            if(!stk.empty())
            {
                nge[i]=stk.top();
            }
            
            stk.push(i);
        }
        
        int ans=0;
        
        for(int i=0;i<nums.size();i++)
        {
            if(nge[i]!=n)
            {
                ans++;
            }
        }
        
        return ans;
    }
};

File: 1996-the-number-of-weak-characters-in-the-game/1996-the-number-of-weak-characters-in-the-game.py
class Solution:
    def numberOfWeakCharacters(self, properties: List[List[int]]) -> int:
        def comp(a,b):
            if a[0]==b[0] : return b[1]-a[1]
            return a[0]-b[0]
        properties.sort(key = cmp_to_key(comp))
        
        res = 0
        # print(properties)
        #------------------------------------------
        arr = [0]*len(properties)
        
        stack = [-1]
        for i in range(len(properties)-1,-1,-1):
            while stack[-1]!=-1 and stack[-1]<=properties[i][1]:
                stack.pop()
            arr[i] = stack[-1]
    
            stack.append(properties[i][1])
        #----------------------------------------------
        # print(arr)
        for i in range(len(properties)):
            if arr[i]!=-1 :
                res+=1
        return res